### ISSUE TITLE
Speed up minikube start

### ISSUE THREAD

**Issue by mcandre** [kubernetes/minikube#1202]
`minikube start` takes 60+ seconds to do its work, which stinks for users who wish to quickly, repeatedly start and tear down several minikube clusters. Can this process be sped up somehow?

**Comment by r2d4** [comment]
Definitely agreed this should be a focus.  Previously, we've been working on the stability of minikube, but now it might make sense to look at performance improvements with starting up.  

It might be useful to document the exact startup process that minikube goes through so we can identify what points can be done in parallel or sped up.

**Comment by aaron-prindle** [comment]
Right now minikube uses docker/libmachine for provisioning VMs and the most of the time is spent provisioning the VM and then waiting for all services to be running.  If you want to try to profile Minikube you can use the MINIKUBE_ENABLE_PROFILING flag here: https://github.com/kubernetes/minikube#minikube-environment-variables

The only way to speed this process up currently would be to run one of the WIP local/no-vm drivers which are much faster as they run everything on the host and not in a VM:
https://github.com/kubernetes/minikube/pull/1173
https://github.com/kubernetes/minikube/pull/1152
NOTE: These currently only support linux and require docker to be installed.

**Comment by dongsupark** [comment]
👍 
I also think it takes longer than expected, until `minikube start` finished. For me it takes mostly up to 120 seconds.
I totally agree that the on-going works mentioned above could really help. However, I'm also wondering, if that's because [localkube service wants network-online.target](https://github.com/kubernetes/minikube/blob/master/pkg/minikube/cluster/commands.go#L43..L44) before it returns. Is that really necessary?

**Comment by aaron-prindle** [comment]
Related: #1277

**Comment by dlorenc** [comment]
We've removed the network-online.target.

**Comment by dfee** [comment]
I found this issue as it's taking minutes for `minikube start` to start for me. I see the issue, "Speed up minikube start" has been closed – is that appropriate?

The full workflow looks like this:
```
minikube delete
minikube start --vm-driver=hyperkit --mount-string=/Users/dfee/code:/code --mount --v=3
minikube stop
minikube start --vm-driver=hyperkit --mount-string=/Users/dfee/code:/code --mount --v=3
```

The restarting surprisingly takes the longest (5+ min)

**Comment by andrew-dinunzio** [comment]
A recent update to Minikube may have made `minikube delete`s much slower too, but I have no idea if it's caused by recent Windows updates since I'm using hyperv on Windows.

**Comment by eliot1785** [comment]
I'd just like to add on to this, I think it should be reopened. For me it's more like a 10 minute startup (the first time I'm running it). I believe that the main issue is that the download of various images is very slow. For me it was as slow as 1/10 MB/s to download the Minikube ISO... I'm in NYC and I have a 150Mbps connection, and I tested separately with SpeedTest.net and I'm getting my promised speed from my provider. Maybe the files are being hosted in a single server far from me.

Specifically, the slowest steps were:

- Downloading Minikube ISO
- Downloading kubeadm v1.13.4 / Downloading kubelet v1.13.4 (I think these might have been in parallel)
- Pulling images required by Kubernetes v1.13.4

**Comment by afbjorklund** [comment]
There's some work ongoing to be able to do those steps ahead-of-time, like "batteries included".

See `--cache-images` and the recent `--download-only` flag that was suggested in #3736

**Comment by eliot1785** [comment]
Thanks for the note. Although I imagine that even if the work is done ahead of time, it will still need to be done again whenever a version update is performed, right?

**Comment by afbjorklund** [comment]
If there is a version update, some of the new components will have to be downloaded. Others not. As in: you can re-use the ISO, even with a new Kubernetes version. And some subcomponents are the same.

**Comment by dgoldssfo** [comment]
Using minikube 0.34.1. I'm working remotely today with a rather slow internet connection. minikube start has been stuck on:

🚜  Pulling images required by Kubernetes v1.13.2 ...

for well over an hour now. 

I realize you don't expect people to be working on such a slow ISP, and maybe I'll have to go somewhere else to work. But how about some diagnostics to let users know what's going on here. Am I 90% done now? 20% done? Or has the thing had to restart 10 times because it lost a connection somewhere along the way? I have no way of telling.

**Comment by afbjorklund** [comment]
It would be nice to have image progress bars, but that requires some new feature in the library used.

It might be doable to make one that shows the number of images though, even if not the actual size...
